python/cugraph-pyg/cugraph_pyg/data/graph_store.py

```diff
@@ -0,0 +1,322 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import cupy
+import cudf
+import pandas
+
+import pylibcugraph
+
+from cugraph.utilities.utils import import_optional, MissingModule
+from cugraph.gnn.comms import cugraph_comms_get_raft_handle
+
+from typing import Union, Optional, List, Dict
+
+
+# Have to use import_optional even though these are required
+# dependencies in order to build properly.
+torch_geometric = import_optional("torch_geometric")
+torch = import_optional("torch")
+tensordict = import_optional("tensordict")
+
+TensorType = Union["torch.Tensor", cupy.ndarray, np.ndarray, cudf.Series, pandas.Series]
+
+
+class GraphStore(
+    object
+    if isinstance(torch_geometric, MissingModule)
+    else torch_geometric.data.GraphStore
+):
+    """
+    This object uses lazy graph creation.  Users can repeatedly call
+    put_edge_index, and the tensors won't be converted into a cuGraph
+    graph until one is needed (i.e. when creating a loader).
+    """
+
+    def __init__(self, is_multi_gpu: bool = False):
+        self.__edge_indices = tensordict.TensorDict({}, batch_size=(2,))
+        self.__sizes = {}
+        self.__graph = None
+        self.__vertex_offsets = None
+        self.__handle = None
+        self.__is_multi_gpu = is_multi_gpu
+
+        super().__init__()
+
+    def _put_edge_index(
+        self,
+        edge_index: "torch_geometric.typing.EdgeTensorType",
+        edge_attr: "torch_geometric.data.EdgeAttr",
+    ) -> bool:
+        if edge_attr.layout != torch_geometric.data.graph_store.EdgeLayout.COO:
+            raise ValueError("Only COO format supported")
+
+        if isinstance(edge_index, (cupy.ndarray, cudf.Series)):
+            edge_index = torch.as_tensor(edge_index, device="cuda")
+        elif isinstance(edge_index, (np.ndarray)):
+            edge_index = torch.as_tensor(edge_index, device="cpu")
+        elif isinstance(edge_index, pandas.Series):
+            edge_index = torch.as_tensor(edge_index.values, device="cpu")
+        elif isinstance(edge_index, cudf.Series):
+            edge_index = torch.as_tensor(edge_index.values, device="cuda")
```

Comment:
What's the reason of creating CPU tensor from numpy and pandas input? Is it because we want to delay data transfer until graph creation? If so, I think we should use `device=edge_index.device` in other cases.

Reply:
I don't think there is an easy way to get the device of a cudf dataframe here

Reply:
I see. It makes sense.